eth/handler.go

```diff
@@ -381,7 +383,7 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		for !unknown && len(headers) < int(query.Amount) && bytes < softResponseLimit && len(headers) < downloader.MaxHeaderFetch {
 			// Retrieve the next header satisfying the query
 			var origin *types.Header
-			if query.Origin.Hash != (common.Hash{}) {
+			if hashMode {
```

Comment:
No, this was actually the fix to the bug :)

If you look a bit down in the code (github cuts it off already), then we use this query object as the iterator itself too (as it already has all the necessary infos the iterator would require too (needed headers, current header, etc)). However, when we're iterating backwards towards the genesis block via hashes, the "finalization" of the iteration points the iterator to the nil hash. However, the previous code assumed that if the hash in the iterator/query is nil, then we're actually doing a number query. So it executed one more step (as the number was valid, `0`).

This fix caches whether the original query had it's origin as a hash or a number, and then we use this caches value to check termination opposed to the current state of the iterator.